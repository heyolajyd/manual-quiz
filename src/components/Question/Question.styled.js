import styled from 'styled-components';
import { breakpoints } from 'styles/global';

export const QuestionContainer = styled.section`
  display: block;
  flex-basis: 200px;
`;

export const Content = styled.div`
  padding: 20px 0;
`;

export const Options = styled.ul`
  list-style-type: none;
  padding: 0;
  margin-bottom: 20px;
`;

export const Item = styled.li`
  margin: 10px 0;
  padding: 10px;
  display: flex;
  align-items: center;
  cursor: pointer;

  &:hover {
    background-color: ${(props) => props.theme.colors.lightCyan};
  }

  @media ${breakpoints.sm} {
    padding: 5px;
  }
`;

export const ItemName = styled.span`
  margin-left: 20px;
`;
