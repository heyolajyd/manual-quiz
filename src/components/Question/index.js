import React from 'react';
import PropTypes from 'prop-types';
import {
  QuestionContainer,
  Content,
  Options,
  Item,
  ItemName,
} from './Question.styled';

function createMarkup(markup) {
  return { __html: markup };
}

const Question = ({ currentQuestion, nextQuestion }) => {
  const { question, options = [] } = currentQuestion;

  const handleNext = (answer) => {
    nextQuestion(answer);
  };

  return (
    <QuestionContainer>
      <Content>{question}</Content>
      <Options>
        {options.map((item) => {
          const optionImage = Object.keys(item)[0];
          const optionName = Object.values(item)[0];

          const selectedOption = optionName.toString();

          return (
            <Item key={optionName} onClick={() => handleNext(selectedOption)}>
              <div dangerouslySetInnerHTML={createMarkup(optionImage)} />
              <ItemName>{optionName}</ItemName>
            </Item>
          );
        })}
      </Options>
    </QuestionContainer>
  );
};

Question.propTypes = {
  currentQuestion: PropTypes.shape({
    question: PropTypes.string,
    options: PropTypes.array,
  }),
  nextQuestion: PropTypes.func,
};

export default Question;
