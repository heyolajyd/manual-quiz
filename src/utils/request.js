/**
 * Util function to get questions
 */
export const getQuestion = async () => {
  try {
    const res = await fetch('./questions.json');
    return await res.json();
  } catch (err) {
    console.log('Failed to parse: ', JSON.stringify(err));
    throw new Error(err);
  }
};
