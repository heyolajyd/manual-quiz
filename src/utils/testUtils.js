import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import { theme } from 'styles/global';

export const withThemeProvider = (component) =>
  render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
