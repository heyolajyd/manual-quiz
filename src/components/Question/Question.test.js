import React from 'react';
import { fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import { withThemeProvider } from 'utils/testUtils';
import Question from './index';

describe('Question', function () {
  beforeEach(() => {
    this.props = {
      currentQuestion: {
        question: 'Test Question',
        options: [{ '<h1>Test Label</h1>': 'test value' }],
      },
      nextQuestion: jest.fn(),
    };
  });

  it('should match snapshot', () => {
    const { container } = withThemeProvider(<Question {...this.props} />);
    expect(container).toMatchSnapshot();
  });

  it('should select and display next question', () => {
    const { getByText } = withThemeProvider(<Question {...this.props} />);

    fireEvent.click(getByText(/test value/i));
    expect(this.props.nextQuestion).toBeCalledWith('test value');
  });
});
