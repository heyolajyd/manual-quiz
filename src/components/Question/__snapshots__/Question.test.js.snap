// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question should match snapshot 1`] = `
.c0 {
  display: block;
  -webkit-flex-basis: 200px;
  -ms-flex-preferred-size: 200px;
  flex-basis: 200px;
}

.c1 {
  padding: 20px 0;
}

.c2 {
  list-style-type: none;
  padding: 0;
  margin-bottom: 20px;
}

.c3 {
  margin: 10px 0;
  padding: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c3:hover {
  background-color: lightCyan;
}

.c4 {
  margin-left: 20px;
}

@media (max-width:576px) {
  .c3 {
    padding: 5px;
  }
}

<div>
  <section
    class="c0"
  >
    <div
      class="c1"
    >
      Test Question
    </div>
    <ul
      class="c2"
    >
      <li
        class="c3"
      >
        <div>
          <h1>
            Test Label
          </h1>
        </div>
        <span
          class="c4"
        >
          test value
        </span>
      </li>
    </ul>
  </section>
</div>
`;
